File: packages/csharp/src/Account/Email.cs
// `isa.Account.Email` — POST /v1/email/enqueue.
//
// Transactional email enqueue. `To` accepts a single address or a list.
// `Attachments` is optional; each entry carries the filename and base64-
// encoded content verbatim — encoding is the caller's responsibility so
// binary payloads (PDFs) do not pay the cost of a UTF-8 round-trip.
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using Isa.Sdk.Zyins;

namespace Isa.Sdk.Account;

/// <summary>One attachment in an email-enqueue request.</summary>
public sealed record EmailAttachment
{
    /// <summary>Attachment filename.</summary>
    [JsonPropertyName("filename")] public string Filename { get; init; } = string.Empty;
    /// <summary>Base64-encoded content. Caller encodes; SDK passes through.</summary>
    [JsonPropertyName("content")] public string Content { get; init; } = string.Empty;
}

/// <summary>Inputs for <see cref="AccountEmail.EnqueueAsync"/>.</summary>
public sealed record EmailEnqueueRequest
{
    /// <summary>Recipient address(es). The single string overload uses
    /// <see cref="ToList"/> with one entry; populate either field, not both.</summary>
    [JsonIgnore] public string? To { get; init; }

    /// <summary>Multiple recipients (server treats as a multi-send).</summary>
    [JsonIgnore] public IReadOnlyList<string>? ToList { get; init; }

    /// <summary>Email subject line.</summary>
    [JsonPropertyName("subject")] public string Subject { get; init; } = string.Empty;

    /// <summary>Body — server treats as HTML when content looks like HTML, else text.</summary>
    [JsonPropertyName("body")] public string Body { get; init; } = string.Empty;

    /// <summary>Optional attachments (each Content is pre-base64-encoded by caller).</summary>
    [JsonPropertyName("attachments")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public IReadOnlyList<EmailAttachment>? Attachments { get; init; }

    /// <summary>Wire field — set by the SDK from <see cref="To"/> or <see cref="ToList"/>.</summary>
    [JsonPropertyName("to")]
    public object ToWire
    {
        get
        {
            var recipients = AccountEmail.NormalizeRecipients(this);
            return ToList is null && recipients.Count == 1 ? recipients[0] : recipients;
        }
    }
}

/// <summary>Result of <see cref="AccountEmail.EnqueueAsync"/>.</summary>
public sealed record EmailEnqueueResult
{
    /// <summary>Normalized to <c>"queued"</c> on success.</summary>
    [JsonPropertyName("status")] public string Status { get; init; } = "queued";
}

/// <summary>`isa.Account.Email` facade.</summary>
public sealed class AccountEmail
{
    private const string Path = "/v1/email/enqueue";
    private readonly AccountContext _ctx;

    internal AccountEmail(AccountContext ctx) => _ctx = ctx;

    /// <summary>Enqueue a transactional email.</summary>
    public Task<EmailEnqueueResult> EnqueueAsync(EmailEnqueueRequest request, CancellationToken ct = default)
    {
        if (request is null) throw new ArgumentNullException(nameof(request));
        var recipients = NormalizeRecipients(request);
        if (recipients.Count == 0)
            throw new ArgumentException("account: email.enqueue requires at least one recipient", nameof(request));
        if (string.IsNullOrEmpty(request.Subject))
            throw new ArgumentException("account: email.enqueue requires a subject", nameof(request));
        if (request.Body is null)
            throw new ArgumentException("account: email.enqueue requires a body", nameof(request));
        var op = _ctx.RequireOp();
        return HttpDispatcher.PostJsonEnvelopeAsync<EmailEnqueueRequest, EmailEnqueueResult>(
            op, Path, request, "email.enqueue", ct);
    }

    internal static IReadOnlyList<string> NormalizeRecipients(EmailEnqueueRequest r)
    {
        if (r.ToList is { Count: > 0 })
        {
            var keep = new List<string>(r.ToList.Count);
            foreach (var x in r.ToList)
            {
                if (!string.IsNullOrEmpty(x)) keep.Add(x);
            }
            return keep;
        }
        if (!string.IsNullOrEmpty(r.To)) return new[] { r.To! };
        return Array.Empty<string>();
    }
}

File: packages/ts/src/account/email.ts
/**
 * `isa.account.email` — `POST /v1/email/enqueue`.
 *
 * Transactional email enqueue. `to` accepts either a single address or a
 * list (server treats `string[]` as a multi-recipient send). `attachment`
 * is optional and carries the filename and base64-encoded content
 * verbatim — encoding is the caller's responsibility so binary payloads
 * (PDFs) do not pay the cost of a UTF-8 round-trip through the SDK.
 *
 * The server response shape is `{ status: 'queued' | '1' }` per the legacy
 * BPP enqueue surface; both values are accepted and normalized to
 * `'queued'`.
 */

import { type AuthContext } from './auth.js';
import { type Transport } from '../zyins/transport.js';
import { fromHttpResponse } from '../zyins/errors.js';
import { deriveIdempotencyKey } from '../zyins/idempotency.js';
import { buildLicenseHMACHeaders } from '../core/index.js';
import { type Clock, systemClock } from '../core/index.js';

const EMAIL_PATH = '/v1/email/enqueue';

/** One attachment in an email-enqueue request. */
export interface EmailAttachment {
  filename: string;
  /** Base64-encoded content. Caller encodes; SDK passes through. */
  content: string;
}

/** Inputs for `account.email.enqueue`. */
export interface EmailEnqueueRequest {
  /** Recipient address(es). */
  to: string | string[];
  /** Email subject line. */
  subject: string;
  /** Body — server treats as HTML when content looks like HTML, else text. */
  body: string;
  /** Optional attachment (`content` is pre-base64-encoded by caller). */
  attachment?: EmailAttachment;
}

export interface EmailEnqueueResult {
  /** Normalized to `'queued'` regardless of which legacy literal the server returns. */
  status: 'queued';
}

export interface EmailContext {
  baseUrl: string;
  auth: AuthContext;
  transport: Transport;
  clock: Clock;
  idempotencyKey?: string;
}

/** Enqueue a transactional email. */
export async function enqueue(
  request: EmailEnqueueRequest,
  ctx: EmailContext,
): Promise<EmailEnqueueResult> {
  if (!request) {
    throw new Error('account: email.enqueue requires a request');
  }
  const recipients = normalizeRecipients(request.to);
  if (recipients.length === 0) {
    throw new Error('account: email.enqueue requires at least one recipient');
  }
  if (typeof request.subject !== 'string') {
    throw new Error('account: email.enqueue requires a subject');
  }
  if (typeof request.body !== 'string') {
    throw new Error('account: email.enqueue requires a body');
  }
  const wire: Record<string, unknown> = {
    to: Array.isArray(request.to) ? recipients : recipients[0],
    subject: request.subject,
    body_html: request.body,
  };
  if (request.attachment) {
    const attachment = request.attachment;
    wire['attachment'] = {
      filename: attachment.filename,
      content_base64: attachment.content,
    };
  }
  const wireBody = JSON.stringify(wire);
  const idempotencyKey =
    ctx.idempotencyKey ??
    (await deriveIdempotencyKey({
      deviceId: ctx.auth.deviceId,
      op: 'email_enqueue',
      body: wireBody,
    }));
  const headers = await buildLicenseHMACHeaders(
    ctx.auth.licenseKey,
    ctx.auth.orderId,
    ctx.auth.email,
    'POST',
    EMAIL_PATH,
    wireBody,
    ctx.auth.deviceId,
    ctx.clock ?? systemClock,
  );
  const response = await ctx.transport({
    url: `${ctx.baseUrl}${EMAIL_PATH}`,
    method: 'POST',
    headers: {
      ...headers,
      'Content-Type': 'application/json',
      Accept: 'application/json',
      'Idempotency-Key': idempotencyKey,
    },
    body: wireBody,
  });
  if (response.status >= 200 && response.status < 300) {
    return { status: 'queued' };
  }
  throw fromHttpResponse(response.status, response.body);
}

function normalizeRecipients(to: string | string[]): string[] {
  if (Array.isArray(to)) {
    return to.filter((entry) => typeof entry === 'string' && entry.length > 0);
  }
  if (typeof to === 'string' && to.length > 0) return [to];
  return [];
}
